---
title: "PTM signatures database (PTMsigDB)"
output:
  html_document:
    toc: no
    toc_float: no
    theme: journal
runtime: shiny
resource_files:
- db/ptm.sig.db.all.v1.9.1.zip
- db/ptm.sig.db.all.v2.0.0.zip
---
<style type="text/css">
.main-container {
  max-width: 1500px;
  margin-left: auto;
  margin-right: auto;
}
</style>

<img src="www/ptmsigdb-logo.png" alt="Drawing" style="width: 250px;"/>

```{r "init", echo=F, message=F}
options(warn = -1)
library(readxl)
library(knitr)
library(glue)
library(magrittr)
library(shiny)
library(DT)
library(kableExtra)

ver.app <- "v0.0.3"

dat.xlsx <- 'data/PTMsigDB_all_sites_v2.0.0.xlsx'
dat.rds <- sub('\\.xlsx', '.rds', dat.xlsx)

ver <- sub('.*_(v.*)\\.xlsx', '\\1' , dat.xlsx)
org.all <- c('human', 'mouse', 'rat')

file.zip <- glue("db/ptm.sig.db.all.{ver}.zip")
file.xlsx <- dat.xlsx
```


```{r "import", echo=F, message=FALSE}
## import
if(!file.exists(dat.rds)){
  
    dat <- lapply(org.all, function(org)read_excel(dat.xlsx, sheet = org) )
    names(dat) <- org.all
    
    ## add links to psp
    ## NOT WORKING
    # dat <- lapply(dat, function(tab){
    #   x <- apply(tab, 1, function(row) {
    #     grpid <- gsub(' ', '', row['site.psp.grpid'])
    #     row['site.psp.grpid'] <- paste0('<a href="https://www.phosphosite.org/siteAction.action?id=', grpid,'" target="_blank_">',  grpid, '</a>')
    #     row
    #   })
    #   t(x)
    # })
    # 
    ## add links to UniProt
    dat <- lapply(dat, function(tab){
      x <- apply(tab, 1, function(row) {
        site.uniprot <- gsub(' ', '', row['site.uniprot'])
        up.id <- sub(';.*', '', site.uniprot)
        row['site.uniprot'] <- paste0('<a href="https://www.uniprot.org/uniprot/', up.id,'#ptm_processing" target="_blank_">', site.uniprot, '</a>')
        row
      })
      t(x)
    })
    
    ## add links to pubmed
    dat <- lapply(dat, function(tab){
      x <- apply(tab, 1, function(row) {
        anno <- gsub(' ', '', row['site.annotation'])
        anno.gn <- sub('^(.*?)\\:.*', '\\1', anno) 
        anno <- sub('^.*?\\:', '', anno)
        anno <- unlist(strsplit(anno, ';')) %>% as.numeric
        if(sum(is.na(anno)) == 0){
          anno <- sapply(anno, function(pmid) paste0('<a href="https://www.ncbi.nlm.nih.gov/pubmed/', pmid, '" target="_blank_">', pmid, '</a>'))
          anno <- paste(anno.gn, paste0(anno, collapse=';'), sep=':')
          row['site.annotation'] <- anno
        }
        row
      })
      t(x)
    })
    ## add links to WikiPathways
    dat <- lapply(dat, function(tab){
      x <- apply(tab, 1, function(row) {
        anno <- gsub(' ', '', row['site.annotation'])
        anno.gn <- sub('^(.*?)\\:.*', '\\1', anno) 
        wp.id <- sub('.*\\:', '', anno)
        anno <- sub('^.*?\\:', '', anno)
        if(grepl('\\.wikipathways\\.', anno)){
            anno <- paste0('<a href="https://www.wikipathways.org/index.php/Pathway:', wp.id, '" target="_blank_">', wp.id, '</a>')
            anno <- paste(anno.gn, paste0(anno, collapse=';'), sep=':')
          row['site.annotation'] <- anno
          
        }
        row
      })
      t(x)
    })
    
    saveRDS(dat, file=dat.rds)
    
} else {
  dat <- readRDS(dat.rds)
}
```

```{r "zip-db", echo = F, message = F, eval = TRUE}
# if(!file.exists(file.zip)){
#   zip(file.zip, files=dir('db/', pattern='.gmt$'))
# }
```

***

The **PTM signatures database (PTMsigDB)** is a collection of modification site-specific signatures of perturbations, kinase activities and signaling pathways curated from literature. PTMsigDB provides the foundation for site-specific **PTM-signature enrichment analysis (PTM-SEA)** of mass spectrometry derived PTM datasets. Please see [Krug et al. 2019](https://www.mcponline.org/content/18/3/576) for details about the curation process of PTMsigDB and PTM-SEA methodology.

### PTM-SEA

Resources to perform PTM Signature Enrichment analysis (PTM-SEA):

```{r "resource-table", echo=F}
tab <- data.frame(
  resource=c('[ssGSEA2.0/PTM-SEA](https://github.com/broadinstitute/ssGSEA2.0)', '[PTM-SEA-GP](https://tinyurl.com/PTM-SEA-GP)'),
  description=c('GitHub page hosting source code of ssGSEA2.0 and PTM-SEA.', '[GenePattern](http://software.broadinstitute.org/cancer/software/genepattern) module to perform PTM-SEA in a web browser.')
)
kable(tab) %>% kable_styling() 
```


***


### PTMsigDB `r ver`

#### Database stats:


```{r "db-stats", echo=F}
nsignat <- sapply(dat, function(x) length(unique(x[, 'signature'])))
ncat <- sapply(dat, function(x) length(unique( sub('-.*', '', x[, 'category']) )))
nsites <- sapply(dat, function(x) length(unique(x[, 'site.uniprot'])))
nproteins <- sapply(dat, function(x) length( unique( sub(';.*','',x[, 'site.uniprot']) )))
ngenes <- sapply(dat, function(x) length( unique( sub('_.*','',x[, 'site.annotation']) )))
kable( data.frame(no.signatures=nsignat,
                  no.categories=ncat,
                  no.sites=nsites,
                  no.proteins=nproteins,
                  no.genes=ngenes) 
       ) %>% kable_styling(bootstrap_options = "striped", full_width = F)

```

#### PTM signature sets:

Each row in the table below corresponds to a PTM site in a specific PTM signature set. Each site is represented by 1) UniProt + residue number; 2) flanking amino acid sequence (+/- 7 amino acids); 3) PhosphoSitePlus site group ids. Please refer to our [GitHub](https://github.com/broadinstitute/ssGSEA2.0#ptmsigdb) page for more details. The site annotation column lists the corresponding gene symbol as well as additional information depending on the signature category. Links to relevant publications (pmid) or WikiPathway entries can be found in column *site.annotation*.

```{r, echo=F}
source('src/shinyDataTable.R')
shinyApp(
  ui = fluidPage(
    column(12, mainPanel(
      tabsetPanel(
        tabPanel('human', shinyDataTableUI(id='human') ),
        tabPanel('mouse', shinyDataTableUI(id='mouse') ),
        tabPanel('rat', shinyDataTableUI(id='rat') )
        )
    ))
  ),
  server = function(input, output) {
    
    callModule(shinyDataTable, id='human', data=dat[['human']])
    callModule(shinyDataTable, id='mouse', data=dat[['mouse']])
    callModule(shinyDataTable, id='rat', data=dat[['rat']])
    },
options = list(height = 750, width=2000)
)
```

<details>
<summary>Signature descriptions</summary>
Signature sets in PTMsigDB are separated into different categories:

+ <code>PERT</code>: Perturbation signatures derived from treatment of cells with perturbagens such as small molecules or growth factors
+ <code>PATH</code>: Signature sets of molecular signaling pathways
+ <code>KINASE</code>: Kinase-substrate signatures
+ <code>DISEASE</code>: Disease-associated signature sets

Signature sets in each category were derived from different sources:

+ <code>PSP</code>: PhosphositePlus ([https://www.phosphosite.org/](https://www.phosphosite.org/))
+ <code>WP</code>: WikiPathways ([https://www.wikipathways.org/](https://www.wikipathways.org/))
+ <code>NP</code>: NetPath ([http://www.netpath.org/](http://www.netpath.org/))
+ <code>P100</code>: signature sets curated from the [LINCS](https://lincsproject.org/) [P100](https://panoramaweb.gs.washington.edu/LINCS/P100/targetedms-showList.view?) assay
+ <code>BI</code>: signature sets defined and curated at the Broad Institute  
+ <code>P100-DIA2</code>: signature sets curated from DIA search of [LINCS](https://lincsproject.org/) [P100](https://panoramaweb.org/Panorama%20Public/2021/LINCS%20PCCSE%20P100%202020/project-begin.view?pageId=Raw%20Data) assay
+ <code>iKiP</code>: signature sets from in vitro Kinase-to-Phosphosite interaction database ([iKiP-DB](https://pubs.acs.org/doi/full/10.1021/acs.jproteome.2c00198))
</details>

***

### Downloads

PTMsigDB signature sets curated for human, mouse and rat can be downloaded below. The zip file contains a total of nine [GMT](https://software.broadinstitute.org/cancer/software/gsea/wiki/index.php/Data_formats#GMT:_Gene_Matrix_Transposed_file_format_.28.2A.gmt.29) files corresponding to three database formats (uniprot, flanking sequence, site group ids) curated for three organims (human, mouse, rat). The Excel file is a representation of the table above. Version v1.8.1 was the initial release of PTMsigDB as described in [Krug et al. 2019](https://www.mcponline.org/content/18/3/576). The v1.8.1 Excel file corresponds to Supplemental Table 1 accompanying the manuscript.


<table>
<tr>
<th>zip</th><th>xlsx</th><th>version</th><th>date</th>
</tr>

<!---v2.0.0--->
<tr>
<td>
```{r, echo=F}
shinyApp(
  ui=fluidPage(fluidRow(downloadLink('download.zip', label = glue("gmt/zip")))),
  server=function(input, output, session){
  output$download.zip <- downloadHandler(filename=function(){file.zip},
       content = function(file){file.copy( file.zip, file)}, contentType = "application/zip"
    )
  }, options = list(height = 20)
)
```
</td>
<td>
```{r, echo=F}
shinyApp(
  ui=fluidPage(fluidRow(downloadLink('download.xlsx', label = glue("xlsx")))),
  server=function(input, output, session){
  output$download.xlsx <- downloadHandler(filename=function(){file.xlsx},
       content = function(file){file.copy( file.xlsx, file)}, contentType = "application/xlsx"
    )
  }, options = list(height = 20)
)
```
</td>
<td>`r ver`</td>
<td>2022-10-25</td>
</tr>

<!---v1.9.1--->
<tr>
<td>
```{r, echo=F}
file.zip.v1.9.1 <- glue("db/ptm.sig.db.all.v1.9.1.zip")
shinyApp(
  ui=fluidPage(fluidRow(downloadLink('download.zip.v1.9.1' , label = glue("gmt/zip")))),
  server=function(input, output, session){
  output$download.zip.v1.9.1 <- downloadHandler(filename=function(){file.zip.v1.9.1},
       content = function(file){file.copy( file.zip.v1.9.1, file)}, contentType = "application/zip"
    )
  }, options = list(height = 20)
)
```
</td>
<td>
```{r, echo=F}
file.xlsx.v1.9.1 <- glue("data/PTMsigDB_all_sites_v1.9.1.xlsx")
shinyApp(
  ui=fluidPage(fluidRow(downloadLink('download.xlsx.v1.9.1', label = glue("xlsx")))),
  server=function(input, output, session){
  output$download.xlsx.v1.9.1  <- downloadHandler(filename=function(){file.xlsx.v1.9.1 },
       content = function(file){file.copy( file.xlsx.v1.9.1 , file)}, contentType = "application/xlsx"
    )
  }, options = list(height = 20)
)
```
</td>
<td>v1.9.1</td>
<td>2020-10-20</td>
</tr>

<!---v1.9.0--->
<tr>
<td>
```{r, echo=F}
file.zip.v1.9.0 <- glue("db/ptm.sig.db.all.v1.9.0.zip")
shinyApp(
  ui=fluidPage(fluidRow(downloadLink('download.zip.v1.9.0' , label = glue("gmt/zip")))),
  server=function(input, output, session){
  output$download.zip.v1.9.0 <- downloadHandler(filename=function(){file.zip.v1.9.0},
       content = function(file){file.copy( file.zip.v1.9.0, file)}, contentType = "application/zip"
    )
  }, options = list(height = 20)
)
```
</td>
<td>
```{r, echo=F}
file.xlsx.v1.9.0 <- glue("data/PTMsigDB_all_sites_v1.9.0.xlsx")
shinyApp(
  ui=fluidPage(fluidRow(downloadLink('download.xlsx.v1.9.0', label = glue("xlsx")))),
  server=function(input, output, session){
  output$download.xlsx.v1.9.0  <- downloadHandler(filename=function(){file.xlsx.v1.9.0 },
       content = function(file){file.copy( file.xlsx.v1.9.0 , file)}, contentType = "application/xlsx"
    )
  }, options = list(height = 20)
)
```
</td>
<td>v1.9.0</td>
<td>2019-03-18</td>
</tr>

<!---v1.8.1--->
<tr>
<td>
```{r, echo=F}
file.zip.v1.8.1 <- glue("db/ptm.sig.db.all.v1.8.1.zip")
shinyApp(
  ui=fluidPage(fluidRow(downloadLink('download.zip.v1.8.1', label = glue("gmt/zip")))),
  server=function(input, output, session){
  output$download.zip.v1.8.1 <- downloadHandler(filename=function(){file.zip.v1.8.1},
       content = function(file){file.copy( file.zip.v1.8.1, file)}, contentType = "application/zip"
    )
  }, options = list(height = 20)
)
```
</td>
<td>
```{r, echo=F}
file.xlsx.v1.8.1 <- glue("data/PTMsigDB_all_sites_v1.8.1.xlsx")
shinyApp(
  ui=fluidPage(fluidRow(downloadLink('download.xlsx.v1.8.1', label = glue("xlsx (Supplemental Table 1)")))),
  server=function(input, output, session){
  output$download.xlsx.v1.8.1 <- downloadHandler(filename=function(){file.xlsx.v1.8.1},
       content = function(file){file.copy( file.xlsx.v1.8.1, file)}, contentType = "application/xlsx"
    )
  }, options = list(height = 20)
)
```
</td>
<td>v1.8.1</td>
<td>2018-12-18</td>
</tr>
</table>


</table>

<br/>

<details>
<summary>Changelog</summary>

* **v2.0.0** Oct 25, 2022
  + Added extended perturbagen DIA signature sets from [LINCS](https://lincsproject.org/) [P100](https://panoramaweb.org/Panorama%20Public/2021/LINCS%20PCCSE%20P100%202020/project-begin.view?pageId=Raw%20Data) assay
  + Added in vitro kinase-to-phosphosite interaction database (iKiP-DB) signature sets from [Mari et al. 2022](https://pubs.acs.org/doi/full/10.1021/acs.jproteome.2c00198)
  + PSP-based signature sets based on Oct 2022 release of PhosphositePlus
  + WP-based signature sets based on WikiPathways <code>gpml</code>-files downloaded on Oct 25, 2022
  + Added signature descriptions and changelog

* **v1.9.1** Oct 20, 2020
  + Added ischemia signature derived from [Mertins et al. 2014](https://pubmed.ncbi.nlm.nih.gov/24719451/)
  + PSP-based signature sets based on August 2020 release of PhosphositePlus
  + WP-based signature sets based on WikiPathways <code>gpml</code>-files downloaded on Aug 10, 2020
  
* **v1.9.0** Mar 06, 2019
  + Added disease-associated signatures from [PhosphositePlus](https://www.phosphosite.org/staticDownloads)
  + PSP-based signature sets based on March 2019 release of PhosphositePlus
  + WP-based signature sets based on WikiPathways <code>gpml</code>-files downloaded on Feb 10, 2019
  
* **v1.8.1** Aug 20, 2018
  + Version released with our [manuscript](https://www.mcponline.org/content/18/3/576.abstract)

</details>

***

<p>
<table width="100%">
<tr>
<td align="left">ptmsigdb-app `r ver.app`</td>
<td align="right"><b>[Proteomics Platform@Broad](https://www.broadinstitute.org/proteomics)</b></td>
</tr>
</table>
</p>

